### ISSUE TITLE
wiki and updater scripts: add version info

### ISSUE THREAD

**Issue by atomGit** [arkenfox/user.js#1042]
i've been asked by several people what version of the user.js to use - it has always been my understanding that vX of the user.js corresponds to the same major version of Firefox, so if one is using FF v78 for example, then they should be using v78 of user.js

if that's the case, then i would propose adding that bit of info to the wiki (either in Overview and/or Implementation) as well as in the updater scripts because i have a hunch many users are ignoring this

here's a potential cut & paste for the wiki...

> The 'arkenfox' user.js version corresponds to the major version of Firefox for which it is intended. For example, user.js version 80 is intended to be used with Firefox version 80.x.

for the updater scripts, shouldn't updating be canceled if the user.js version exceeds the Firefox version?

**Comment by Synkky** [comment]
Canceling it just because it exceeds the version of FF doesn't much sense to me. Users should be free to do whatever they want with user.js, and that includes using it "incorrectly".
I'm all for providing that info in both wiki and updater scripts though (even though I personally find it intuitive enough that it doesn't need to be explicitly said)

**Comment by atomGit** [comment]
i agree that users should be free to do as they wish and i'm not suggesting forceful canceling of the update in the event of version mismatch, but perhaps soft-canceling with an option to override (of course they can always manually download it as well)

as the author of a couple of Firefox config guides, it has become apparent to me that there's a significant number of people that are using this js who are not really technically inclined, yet are rightfully concerned about privacy - my thinking regarding the update scripts accounts for this audience (dummy-proof)

my concern is that a version mismatch could break functionality or activate prefs that are available, but not ready for prime-time and this could potentially lead to leakage

**Comment by Synkky** [comment]
I agree on all points.
Let's wait for what Pants thinks of this.

**Comment by Thorin-Oakenpants** [comment]
> intuitive enough that it doesn't need to be explicitly said

Exactly

> Canceling it just because it exceeds the version of FF doesn't much sense to me

Not going to ever block an update, sorry. Not really interested in a second confirmation prompt after an alert either

> Users should be free to do whatever they want with user.js, and that includes using it "incorrectly".

There is no "incorrect". The user.js allows ESR users to flip the deprecated section, and there is info throughout the js file for ESR users: such as the RFP alts section, deprecated section, it was in the readme at the top, we add default version info for ESR users as well: in other words, we actually enable ESR users to use current

A year is a long time between ESR versions. Most changes shouldn't affect ESR users: e.g. new prefs, default values are notated, and deprecated ones are covered. Cleaned up sections, better test links, better info, etc are all immaterial, but useful for ESR

The problem is if we flip something like HTTPS-Only mode when it's ready and suddenly ESR users get a broken feature: and we'll just handle that like we always have (although I'm struggling a little to think of a similar situation, but I don't think we've ever really held off flipping something just because of ESR)

As for the wiki .. sheesh, not more crap to add, please. I'd like to actually go through it and reduce the content by 50%

**Comment by Thorin-Oakenpants** [comment]
bump @earthlng your thoughts please. My thoughts are to close the ticket and make zero changes

**Comment by earthlng** [comment]
there's clearly an issue or @atomGit wouldn't have opened this ticket.
You've now removed the line in the user.js that said to always use the latest master and maybe that's enough to make things less confusing and a bit more intuitive ie use the version that matches your FF version.

> in other words, we actually enable ESR users to use current

ESR is a bit more tricky. There's only so much we can do for ESR users. For example I think RFP in ESR doesn't have the font-vis limitation yet, right? And we recently enabled doc-fonts in the master which probably means we made font-FPing much more problematic for ESR users if they use the master.

As for the updater scripts, it's not easily possible to link a profile to a certain FF installation. At least definitely not on Windows where users could have one version installed in their C drive and numerous other portable FF installations elsewhere.
The dev-edition also installs into a separate directory and can be used in parallel with a stable or ESR installation. I assume that's the same on Linux/Mac too.
The only semi-viable thing I could see doing is parse the prefs.js for the FF version but that can cause some problems too.
F.e. if someone on Release day closes their FF, updates it from the repos or via a setup file and then runs our updater script **before** starting FF again -> the prefs.js would still have the old version.

I think the best we can do, is to hold pref changes for the next version in a PR until a day or 2 after that version is released.
Current example would be the 2 `browser.search.geoSpecificDefaults` removed prefs in FF82.
That and maybe add something to the wiki or Readme, IDK.

**Comment by Thorin-Oakenpants** [comment]
> You've now removed the line in the user.js that said to always use the latest master and maybe that's enough to make things less confusing and a bit more intuitive ie use the version that matches your FF version

That's why I removed it

Yeah. font vis vs font docs is definitely one. HTTPS-Only mode would be another messy one. There's no way we can hold things back for up to a year for ESR users

I actually think we should remove all/most lines/mentions of ESR (e.g. under RFP alts, in the deprecated, and the releases under readme) and state under releases: ESR users are expected to use the correct release or something - end of story

note: we still keep the deprecated of course, just not say ESR78.x instead say FF78

**Comment by Thorin-Oakenpants** [comment]
> end of story

done. now we do not have to worry about ESR. We add recent default info changes for the info factor, we can flip things without worrying about backporting upstream, or dealing with other quirks

The line `Use the correct release that matches your Firefox version` should satisfy this original issue (updaters are too problematic see earthlng)

**Comment by earthlng** [comment]
> remove all/most lines/mentions of ESR

nah, I don't like that at all. There's only 1 or maybe a few things that ideally should to be changed if someone wants to use the master js with ESR. I'd prefer to create a new sticky for ESR users with some infos and maybe a code block with some overrides, OR just keep everything as is.

People using ESR probably do so because they don't want to bother with all the changes all the time and so they probably already use the 78 release anyways or if they're using the master they probably/hopefully follow the changelogs and make the few necessary changes in their overrides. And the few hints and notes we have for them help with that.

**Comment by rusty-snake** [comment]
>  I'd prefer to create a new sticky for ESR users with some infos and maybe a code block with some overrides

Alternative you could create a separate ESR branch/file. However this would mean doing a lot of things twice (broken links, ...). IMHO the best would be to put such a code-block in a own file (e.g. user-esr.js) and the updater-scripts pull user.js and user-esr.js for ESR users. So the final user.js in the firefox-profile would be user.js+user-esr.js+user-overrides.js.

**Comment by Thorin-Oakenpants** [comment]
I don't see a problem. I can remove that line as well, since there are already sub-numbers after that. We of course keep that section so people can use prefsCleaner or reset manually

Right: So you put back all the stuff I took out, and then want to add *more* stuff? Sure, I get it, we accommodated ESR for ages

> There's only 1 or maybe a few things that ideally should to be changed if someone wants to use the master js with ESR

And what do they gain by not using the 78 release in comparison?

**Comment by Thorin-Oakenpants** [comment]
> Alternative you could create a separate ESR branch/file.

apologies if I get a bit blue here, this is not aimed at the poster... it's just to make this crystal clear .... FUCK NO

**Comment by earthlng** [comment]
> And what do they gain by not using the 78 release in comparison?

maybe not very much at the moment but we're only 4 versions in since the last ESR. Some upcoming changes we'll maybe make could help them too, fe think of the time when we changed the cookie pref from to 2 to 1 to reduce breakage.

> I can remove that line as well

at least one of the updater script relies on that line to re-enable the ESR section

**Comment by atomGit** [comment]
apparently warning the user about a version mismatch between the user.js and ff version is off the table, and i get that, but i'll mention this anyway...

@earthlng said..

> As for the updater scripts, it's not easily possible to link a profile to a certain FF installation. [...] The only semi-viable thing I could see doing is parse the prefs.js for the FF version but that can cause some problems too.

maybe i'm not understanding, but on Linux you can `firefox -v`(bash: `ffvn="$(firefox -v)"` ) - i assume this works on Windows as well

**Comment by rusty-snake** [comment]
@atomGit the issue with `firefox -v` is that a user can have additional installations (e.g. nightly) in it's home. So you can not be sure that the firefox binary in $PATH is that used with this profile. However it is possible to `grep "extensions.lastAppVersion" prefs.js` or `grep "LastVersion" compatibility.ini`.

**Comment by Synkky** [comment]
This discussion has taken so many turns I don't even know what to say.
I understand earthlng's point of view, but at the same time, I'm pretty sure there's no reason why an ESR user would ever use master unless they were after some tweaking thrills.
It's true that currently it only takes a few tweaks to get it pretty similar (e.g. font vis vs font docs / HTTPS-Only whenever that gets flipped), but there's no way to actually predict whether it'll be possible to keep those tweaks to a minimum in the future (depending on the context and aftereffects of current prefs or new prefs in the future) and you have to admit that assuring that those tweaks remain as minimal as possible in the future **just because** someone may want to use master on an ESR release is a blatant waste of time **imo**.

Please, let's just go with the simplest solution:
- Remove all ESR stuff from user.js
- Make a dedicated ESR wiki page to explain how this works for ESR users **in big bold letters** (e.g. using only the user.js version that matches their current version) because people have to read the wiki to begin with, otherwise they won't be able to use user.js properly
- (optional) Change the updater scripts in a way they'll detect ESR and either download the proper version (and its possible .x patch versions, if applicable, although patch versions rarely modify prefs) or do nothing and just exit with an info message (e.g. download only the required version manually or whatever)

**Comment by Thorin-Oakenpants** [comment]
It's not so much a question of can we do this, because we can (but it does become just another thing to think about and cater for): it's a question why should we support ESR using the current master

ESR users choose to use a version that only gets security updates. They know they will be on the same version for a year. Regardless of what we do in the user.js, when they come to upgrade to the next ESR, they are in for a year's worth of changes: both from Mozilla and from the user.js. The user.js mitigates it somewhat: we have the deprecated items listed, and we have the removed items scratchpad, and we have changelogs (all 12 or 13 of them), and we have version numbers for new prefs (so the user can search for `[FF79` and `[FF8` etc) - not ideal, but I don't think there's anything else we can do

---

But lets say an ESR user likes to update to master all the time: what scenarios can happen
- removals
   - e.g. something that is still used by ESR. We're pretty careful about this, but it could happen. But unless the user manually resets it or runs the removed script, it wouldn't affect them
- deprecated
   - user should switch the appropriate deprecated section back on: but this isn't necessarily required, just best practice. If the items have already been set, then unless they run prefsCleaner or manually reset them, then they don't change
- updated info
   - things like revamped sections, new sections, better references, updated stats, default value info, etc: nice to have but not strictly essential

prefs: there will be an overlap of when a pref is added by Mozilla and when we do something with it in the user.js: e.g. add/remove/change-active-status/flip-value

- new inactive prefs
   - zero affect
- new to ESR+1 active prefs
   - zero affect
- prefs that overlap ESR + current release
   - either adding them, or flipping them, or making them active/inactive
   - ✅ they can help
      - e.g. we comment something out for breakage (see cookie value, language section, clipboardevents, mathml, svg etc changes: lots of these: but we should be fairly stable by now) or for functionality (see `urlbar.suggest*` getting made inactive) - I mean we're always improving and that includes existing prefs
      - e.g. we discover an old pref (it happens: e.g. I have a webgl one that looks very likely)
   - ❌ they can conflict (examples only)
      - we add/turn on blocking insecure downloads
      - we turn on HTTPS-Only mode
      - we turn on DoH
      - we switch to dFPI
      - we enable next gen local storage

We don't know what we're going to add or change in the future

I am OK with accommodating ESR, but I want to do it differently than before. I am not interested in adding more wiki stuff (I want to clean that out and super simplfy it, and maybe restyle it)

What we need is a central spot in the user.js to mention ESR items: and the place to do that is under releases, IMO. I'll do a rough idea of what I mean in the next post

**Comment by Thorin-Oakenpants** [comment]
in that list above where it conflicts, this can also be a bit misleading: if we make something inactive, the user would still have to run prefsCleaner or manually reset it: so something like 1401 might not have has an effect. But it also cuts both ways, e.g. we relax something for less breakage

**Comment by Thorin-Oakenpants** [comment]
what was taken out (so we can start from scratch)

```
readme
 * EACH RELEASE check:
         - 4600s: reset prefs made redundant due to privacy.resistFingerprinting (RPF)
                  or enable them as an alternative to RFP (or some of them for ESR users)
         - 9999s: reset deprecated prefs in about:config or enable the relevant section for ESR

1007
 * [SETUP-WEB] ESR78: playback might break on subsequent loading (1650281)

RFP Alts section header
   * ESR RFP users:
       Reset those *up to and including* your version. Add those *after* your version
       as active prefs in your overrides. This is assuming that the patch wasn't also
       backported to Firefox ESR. Backporting RFP patches to ESR is rare.

depecated note
  // [NOTE] replace the * with a slash in the line above to re-enable them
```

We can put back the 1007 line and the note on how to flip the deprecated section.  I also want to clean up the RFP Alts section header... probably almost leave it empty: the shit-to-check every release should be in one place

anyway, here's what I have in mind (super rough)

```
* RELEASES: https://github.com/arkenfox/user.js/releases

  * VERSIONS
    - It is best to use the arkenfox release that matches your Firefox version
  * ESR78
    - If you are not using arkenfox v78... this is not a definitive list
      - 1401: document fonts is inactive
      - 4600: rfp alternatives
      - 9999: switch the appropriate deprecated section back on
  * non-ESR
    - run the prefsCleaner or reset deprecated prefs (9999s) and prefs made redundant by RPF (4600s)
    - re-enable section 4600 if you don't use RFP
```

That gives us a place to add things like `1244: HTTPS-Only mode is enabled: this is broken in ESR`: Honestly, I don't see it getting too big over each cycle, since we only need to list things that make them worse off

**Comment by Thorin-Oakenpants** [comment]
bump ^^ if you're happy with that approach, which is way better than what we had, then let me know and I will do a PR
- it covers the issue of stating the obvious about versions... so @atomGit won't be TRiGGeReD anymore 😁 
- it centralizes the info
- it allows us to add specific items like doc fonts and HTTPS-Only mode
- ideas welcome
- drafts welcome (I hate writing these things: trying to keep them really short but cover all the bases)

**Comment by earthlng** [comment]
I like the centralized approach but I still think it'd be better in a sticky issue because it's more noticeable when we make changes.
Also allows us to provide a recommended overrides code-block for example. And then in the user.js we can just link to that issue.

If you like it better if it's all in the user.js then go ahead and make the changes you want

**Comment by Thorin-Oakenpants** [comment]
people don't read the wiki, people don't read issues

if it's in the user.js, then if someone says .. why this, why that, wha wha wambulance ... I can say, well, did you read the actual user.js

**Comment by earthlng** [comment]
you can still say that because the user.js will have a link to the issue :)

**Comment by Thorin-Oakenpants** [comment]
true, but people still don't do it (go read stuff) - remove a barrier (non clickable link) and provide the actual answer right there on the spot

We could have both - but only if you promise to maintain the sticky issue

**Comment by Thorin-Oakenpants** [comment]
actually, lets do that: you create a new issue for a sticky and we'll change it to something like this

```
* RELEASES: https://github.com/arkenfox/user.js/releases

  * VERSIONS
    - It is advised you use the arkenfox release that matches your Firefox version
  * EACH RELEASE
    - run the prefsCleaner or reset deprecated prefs (9999s) and prefs made redundant by RPF (4600s)
    - re-enable section 4600 if you don't use RFP
  * ESR78... if you are not using the recommended arkenfox v78...
    - do NOT do the above... instead read [link]

```

I'm more than happy to have less in the user.js and not encourage ESR to use current master (since I was willing to actually to remove it altogether), and to delegate it's maintenance to someone else .. sweeeet

**Comment by Thorin-Oakenpants** [comment]
^^ I'm sold on it. Less work for me 🍻

**Comment by handshake900** [comment]
Just one question: FFv82 is out today. Current user.js is v81.

Do I update FF today, or wait for the next version of the user.js?

Respect and regards..

**Comment by atomGit** [comment]
the user.js is 82 alpha

and yes, you should always keep FF up to date, even if the user.js is lagging ... which i don't think it ever has because pants, earthing and the boys aren't slackers :)

**Comment by handshake900** [comment]
Thank you for the response.

I really appreciate the work you guys do.

**Comment by handshake900** [comment]
Pardon me, two more:

I understand FPI. With that said, are any container extensions necessary?

Always on VPN. Always. Encrypted DNS: yes or no.

Thanks again.

**Comment by atomGit** [comment]
in my *personal* opinion (and i'm not the sharpest knife in the drawer), i find containers unnecessary - i think FPI is more transparent (it is for me anyway), requires less fiddling, plus it's built-in, and it doesn't break nav'ing through history

there is of course more to the story and your needs may be very different

i don't know enough about DoH to comment on it, other than to say that i would NOT be using CloudFlare or any other privacy-hating entity as a DNS provider (i believe CF is the Firefox default) - you could also run DNS through your VPN (any decent VPN services offers this)

**Comment by Thorin-Oakenpants** [comment]
> even if the user.js is lagging 

well, we can't put out a beta until after the release lands, and it can take a few days .. but it's nothing to worry about: I mean you've been using the same setup for the last four to six weeks ... you'll survive .. it's not like FF is going to regress in security or privacy ... update Firefox when you want to, within reason: I always have to wait until portablesapps put theirs out (I could update from the menu, but I prefer to re-unpack the full bundle from portableapps, not a maintenance download from mozilla)

PS: I think the longest it took was about 4 weeks (because I was deliberately not doing anything to see what would happen) .. but other than that we've always been inside a week (from memory)

**Comment by handshake900** [comment]
I appreciate the response. My routine is simple/stupid: New profile-drop in the user.js-add UBO, UMatrix and HTTPZ. Rinse and repeat for any profile then ditch the old. Everything backed up and ready to go. Compartmentalized profiles.

Thanks again.

**Comment by handshake900** [comment]
In regard to the uMatrix departed, extensions and the user.js

I’ve always used the ghacks/arkenfox user.js as is, even though it’s “a template”. No issues with that setup period.

Finally let go of uMatrix for now. I’ll wait to see what happens. I believe the extension makes it way quicker to unbreak a site than a hardened uBlock-O setup. However, I use the wiki and the 12bytes.org instructions for reference and have a pretty good understanding of it now.

Having said this, I am curious to know what privacy/security extensions you use. I prefer to use as little browser-condoms as I can. And of course, FPI does so much on it’s own, along with everything else modified via the user.js.

Best regards

**Comment by atomGit** [comment]
> I use the wiki and the 12bytes.org instructions for reference  ...

if you're using my uBO guide, revisit it again tomorrow because i'm updating it

i too am attempting, once again, to rely on uBO in advanced mode without uM

**Comment by handshake900** [comment]
Okay, sounds good.

Just out of curiosity, in regard to the 12bytes extension: does it notify the user of changes similarly to the linux script?

Thanks.

**Comment by atomGit** [comment]
not sure what you mean by a '12bytes extension' - if you're referring to my [user.js-notify.sh](https://codeberg.org/12bytes.org/Firefox-user.js-supplement/src/branch/master/misc/user.js-notify.sh) script, it does not replace the 'arkenfox' updater script - it only checks if a new version of user.js and, optionally, my user-overrides.js is available - you still need the updater script to grab user.js

the purpose of my script is to add it to your auto-start programs (can be done in different ways) so it checks automatically and displays a desktop notification with local and remote version info, so in other words you don't need to check manually by running the updater script

**Comment by handshake900** [comment]
Terrible wording on my part. Today is the day after Friday night :)

I was asking if the 12bytes add-on did something similar to the script as far as notifications to changes. Again, I may have just failed at using words. Damn.

Thanks

**Comment by atomGit** [comment]
> Today is the day after Friday night

you're tellin' me o_O

so did i answer your question?

**Comment by handshake900** [comment]
Yep. Got it.

Thanks

**Comment by atomGit** [comment]
@handshake900 - i'm done with my uBO guide - for changes see the [Revisions](https://12bytes.org/articles/tech/firefox/ublock-origin-suggested-settings/#Revisions) section

**Comment by handshake900** [comment]
Nice. Cleaned it up and is concise. I agree that globally blocking JS is the way to go.

That being said, if someone forks uM and it gets vetted by you guys, I would do it in a heartbeat:) Everyone thought it was hard to use. I felt it was easier, as you could see it all(mostly)

Thanks again and good work
